src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,180 @@
+  rcases hp2 with ⟨p2a, ⟨hp2a, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv1p, hv2p, vsub_vadd_eq_vsub_sub V (v1 +ᵥ p1a), vadd_vsub_assoc, add_comm, add_sub_assoc],
+  have hv1v2 : v1 - v2 ∈ (vector_span k V s).carrier,
+  { apply (vector_span k V s).add hv1,
+    rw ←neg_one_smul k v2,
+    exact (vector_span k V s).smul (-1 : k) hv2 },
+  refine (vector_span k V s).add _ hv1v2,
+  unfold vector_span,
+  change p1a -ᵥ p2a ∈ submodule.span k (vsub_set V s),
+  have hp1p2 : p1a -ᵥ p2a ∈ vsub_set V s, { use [p1a, hp1a, p2a, hp2a] },
+  have hp1p2s : vsub_set V s ⊆ submodule.span k (vsub_set V s) := submodule.subset_span,
+  apply set.mem_of_mem_of_subset hp1p2 hp1p2s
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+section affine_subspace
+
+variables (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
+(carrier : set P)
+(direction : subspace k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction.carrier → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def univ_affine_subspace : affine_subspace k V P :=
+{ carrier := set.univ,
+  direction := submodule.span k set.univ,
+  nonempty := set.nonempty_iff_univ_nonempty.1 S.nonempty,
+  add := λ p v hp hv, set.mem_univ _,
+  sub := begin
+    intros p1 p2 hp1 hp2,
+    apply set.mem_bInter,
+    intros x hx,
+    rw set.mem_set_of_eq at hx,
+    exact set.mem_of_mem_of_subset (set.mem_univ _) hx
+  end }
+
+instance : inhabited (affine_subspace k V P) := ⟨univ_affine_subspace k V P⟩
+
+/-- The affine span of a nonempty set of points is the smallest affine
+subspace containing those points. (Actually defined here in terms of
+spans in vector spaces.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
+  direction := vector_span k V s,
+  nonempty := span_points_nonempty_of_nonempty k V h,
+  add := λ p v hp hv, vadd_mem_span_points_of_mem_span_points_of_mem_vector_span k V hp hv,
+  sub := λ p1 p2 hp1 hp2, vsub_mem_vector_span_of_mem_span_points_of_mem_span_points k V hp1 hp2 }
+
+end affine_subspace
+
+section affine_map
+
+variables  (k : Type*) (V1 : Type*) (P1 : Type*) (V2 : Type*) (P2 : Type*)
+    (V3 : Type*) (P3 : Type*) [field k]
+    [add_comm_group V1] [vector_space k V1] [affine_space k V1 P1]
+    [add_comm_group V2] [vector_space k V2] [affine_space k V2 P2]
+    [add_comm_group V3] [vector_space k V3] [affine_space k V3 P3]
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
```

Comment:
I think it would be better to `extends linear_map k V1 V2`.

Reply:
That would mean `affine_map.to_fun` is the `to_fun` of the `linear_map`, and the function from points to points is called something else. Is that's what intended?

Reply:
That doesn't sound so good.